plugins/kubernetes/app/models/watchers/base_watcher.rb

```diff
@@ -8,18 +8,24 @@ class BaseWatcher
     def initialize(watch_stream, log: true)
       @watch_stream, @log, = watch_stream, log
       @start_time = Time.now
+      @enabled = false
     end
 
     def start_watching(&block)
+      @original_thread_id = Thread.current.object_id
+      @enabled = true
       @watch_stream.each do |notice|
         handle_notice(notice, &block)
+        break if stop_watching?
       end
+      cleanup
     end
 
     def stop_watching
-      if @watch_stream
-        @watch_stream.finish
-        @watch_stream = nil
+      if @enabled
+        @enabled = false
+        # close the socket and stop receiving updates if called from a different thread than the original
+        cleanup if Thread.current.object_id != @original_thread_id
```

Comment:
I'm still puzzled about how can this be happening, I mean, reaching this point from a different Thread. If all this objects (ReleaseGroup, Release, DeployService and PodWatcher) get created within the scope of a request, and then a new Thread is launched from this request to watch for the state of this objects, with no state being shared across different Threads (apparently), how can we end up with this scenario? 

I'm just concerned there's something else we're missing in here that could lead us to some weird and inconsistent state. :(